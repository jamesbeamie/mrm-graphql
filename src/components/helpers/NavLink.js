import React from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';
import Link from 'react-toolbox/lib/link';

/**
 * Handles redirections upon click
 * of any component on the NavLink
 *
 * @param {object} event
 * @param {object} history
 * @param {object} to
 *
 * @returns {JSX}
 */
export const handleClick = (event, history, to) => {
  event.preventDefault();

  history.push(typeof to === 'object' ? to.pathname : to);
};

const NavLink = ({
  to, exact, strict, ...rest
}) => (
  <Route
    path={to}
    exact={exact}
    strict={strict}
    children={({ history, match }) => (
      <Link
        {...rest}
        active={!!match}
        onClick={event => handleClick(event, history, to)}
      />
    )}
  />
);

NavLink.propTypes = {
  to: PropTypes.oneOfType([
    PropTypes.shape({
      pathname: PropTypes.string.isRequired,
    }),
    PropTypes.string,
  ]).isRequired,
  exact: PropTypes.bool,
  strict: PropTypes.bool,
};

NavLink.defaultProps = {
  exact: false,
  strict: false,
};

export default NavLink;
