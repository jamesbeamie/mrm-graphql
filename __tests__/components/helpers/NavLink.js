import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import NavLink, { handleClick } from '../../../src/components/helpers/NavLink';

describe('NavLink', () => {
  const preventDefault = jest.fn();
  const history = { push: jest.fn() };
  const to = '/home';
  const toObject = { pathname: '/index' };
  const wrapperCode = (
    <MemoryRouter keyLength={0}>
      <NavLink to="/" exact strict />
    </MemoryRouter>
  );
  const wrapper = mount(wrapperCode);

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should respond to click events properly', () => {
    const onClick = () => wrapper.simulate('click', { preventDefault }, history, to);
    expect(onClick).not.toThrow();
    expect(wrapper).toMatchSnapshot();
  });

  it('handleClick method should be called with the right props', () => {
    handleClick({ preventDefault }, history, to);
    expect(preventDefault).toBeCalled();
    expect(history.push).toBeCalled();
    expect(history.push).toBeCalledWith(to);

    handleClick({ preventDefault }, history, toObject);
    expect(preventDefault).toBeCalled();
    expect(history.push).toBeCalled();
    expect(history.push).toBeCalledWith(toObject.pathname);
  });
});
