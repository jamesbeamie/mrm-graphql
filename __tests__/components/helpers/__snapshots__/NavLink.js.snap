// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLink should render correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NavLink
      exact={true}
      strict={true}
      to="/"
    >
      <Route
        exact={true}
        path="/"
        strict={true}
      >
        <ThemedLink
          active={true}
          className=""
          composeTheme="deeply"
          mapThemrProps={[Function]}
          onClick={[Function]}
        >
          <Link
            active={true}
            className=""
            onClick={[Function]}
            theme={Object {}}
          >
            <a
              className="undefined"
              data-react-toolbox="link"
              onClick={[Function]}
            />
          </Link>
        </ThemedLink>
      </Route>
    </NavLink>
  </Router>
</MemoryRouter>
`;

exports[`NavLink should respond to click events properly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "PUSH",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 1,
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NavLink
      exact={true}
      strict={true}
      to="/"
    >
      <Route
        exact={true}
        path="/"
        strict={true}
      >
        <ThemedLink
          active={true}
          className=""
          composeTheme="deeply"
          mapThemrProps={[Function]}
          onClick={[Function]}
        >
          <Link
            active={true}
            className=""
            onClick={[Function]}
            theme={Object {}}
          >
            <a
              className="undefined"
              data-react-toolbox="link"
              onClick={[Function]}
            />
          </Link>
        </ThemedLink>
      </Route>
    </NavLink>
  </Router>
</MemoryRouter>
`;
